pkg/probes/probes.go

```diff
@@ -0,0 +1,252 @@
+	ConfigNetfilterXtMatchBpf    KernelParam `json:"CONFIG_NETFILTER_XT_MATCH_BPF"`
+	ConfigNetActBpf              KernelParam `json:"CONFIG_NET_ACT_BPF"`
+	ConfigNetClsAct              KernelParam `json:"CONFIG_NET_CLS_ACT"`
+	ConfigNetClsBpf              KernelParam `json:"CONFIG_NET_CLS_BPF"`
+	ConfigNetSchIngress          KernelParam `json:"CONFIG_NET_SCH_INGRESS"`
+	ConfigSockCgroupData         KernelParam `json:"CONFIG_SOCK_CGROUP_DATA"`
+	ConfigTestBpf                KernelParam `json:"CONFIG_TEST_BPF"`
+	ConfigTracing                KernelParam `json:"CONFIG_TRACING"`
+	ConfigUprobeEvents           KernelParam `json:"CONFIG_UPROBE_EVENTS"`
+	ConfigXdpSockets             KernelParam `json:"CONFIG_XDP_SOCKETS"`
+	ConfigXfrm                   KernelParam `json:"CONFIG_XFRM"`
+	BpfJitEnable                 int         `json:"bpf_jit_enable"`
+	BpfJitHarden                 int         `json:"bpf_jit_harden"`
+	BpfJitKallsyms               int         `json:"bpf_jit_kallsyms"`
+	BpfJitLimit                  int         `json:"bpf_jit_limit"`
+	UnprivilegedBpfDisabled      int         `json:"unprivileged_bpf_disabled"`
+}
+
+// Features represents information about BPF features returned by libbpf and
+// bpftool.
+type Features struct {
+	Helpers       `json:"helpers"`
+	MapTypes      `json:"map_types"`
+	ProgramTypes  `json:"program_types"`
+	SyscallConfig `json:"syscall_config"`
+	SystemConfig  `json:"system_config"`
+}
+
+// probeKernelConfig checks required and recommended kernel configuration
+// parameters.
+func probeKernelConfig(systemConfig SystemConfig) error {
+	if !systemConfig.ConfigBpf.Enabled() {
+		return fmt.Errorf("CONFIG_BPF kernel parameter is required")
+	}
+	if !systemConfig.ConfigBpfSyscall.Enabled() {
+		return fmt.Errorf("CONFIG_BPF_SYSCALL kernel parameter is required")
+	}
+	if !systemConfig.ConfigNetSchIngress.EnabledOrModule() {
+		return fmt.Errorf("CONFIG_NET_SCH_INGRESS kernel parameter (or module) is required")
+	}
+	if !systemConfig.ConfigNetClsBpf.EnabledOrModule() {
+		return fmt.Errorf("CONFIG_NET_CLS_BPF kernel parameter (or module) is required")
+	}
+	if !systemConfig.ConfigNetClsAct.Enabled() {
+		return fmt.Errorf("CONFIG_NET_CLS_ACT kernel parameter is required")
+	}
+	if !systemConfig.ConfigBpfJit.Enabled() {
+		return fmt.Errorf("CONFIG_BPF_JIR kernel parameter is required")
+	}
+	if !systemConfig.ConfigHaveEbpfJit.Enabled() {
+		return fmt.Errorf("CONFIG_HAVE_EBPF_JIT kernel parameter is required")
+	}
+	if !systemConfig.ConfigCgroupBpf.Enabled() {
+		log.Warning("CONFIG_CGROUP_BPF kernel parameter is not in kernel configuration")
+	}
+	if !systemConfig.ConfigLwtunnelBpf.Enabled() {
+		log.Warning("CONFIG_LWTUNNEL_BPF kernel parameter is not in kernel configuration")
+	}
+	if !systemConfig.ConfigBpfEvents.Enabled() {
+		log.Warning("CONFIG_BPF_EVENTS kernel parameter is not in kernel configuration")
+	}
+	return nil
+}
+
+func probes(features Features) error {
```

Comment:
I don't think this function is really necessary as you are calling `probeKernelConfig` passing the same parameters and returning the same structures.

Reply:
In future we will have more probes, so I expect that function to look like:

```
func probes(features Features) error {
	if err := probeHelpers(features.Helpers); err != nil {
		return err
	}
	if err := probeMapTypes(features.MapTypes); err != nil {
		return err
	}
	if err := probeKernelConfig(features.SystemConfig); err != nil {
		return err
	}
	return nil
}
```

I want to create follow up PR(s) with more probes right after this one gets merged. I didn't want to do everything at once here, because this PR is already quite big itself.